src/jaxsim/rbda/contacts/relaxed_rigid.py

```diff
@@ -390,124 +426,153 @@ def continuing_criterion(carry):
                 a_ref: offset acceleration in constraint frame
-                R: regularization matrix
                 K: computed stiffness
                 D: computed damping
             """
-            position = jnp.zeros(shape=(3,)).at[2].set(penetration)
 
-            imp_x = jnp.abs(position) / width
-            imp_a = (1.0 / jnp.power(mid, p - 1)) * jnp.power(imp_x, p)
+            imp_x = jnp.abs(pos) / width
 
+            imp_a = (1.0 / jnp.power(mid, p - 1)) * jnp.power(imp_x, p)
             imp_b = 1 - (1.0 / jnp.power(1 - mid, p - 1)) * jnp.power(1 - imp_x, p)
-
             imp_y = jnp.where(imp_x < mid, imp_a, imp_b)
 
-            imp = jnp.clip(ξ_min + imp_y * (ξ_max - ξ_min), ξ_min, ξ_max)
-            imp = jnp.atleast_1d(jnp.where(imp_x > 1.0, ξ_max, imp))
+            # Compute the impedance.
+            ξ = ξ_min + imp_y * (ξ_max - ξ_min)
+            ξ = jnp.clip(ξ, ξ_min, ξ_max)
+            ξ = jnp.where(imp_x > 1.0, ξ_max, ξ)
+
+            # Compute the spring and damper parameters during runtime from the
+            # impedance and other contact parameters.
+            K = 1 / (ξ_max * Ω * ζ) ** 2
+            D = 2 / (ξ_max * Ω)
 
-            # When passing negative values, K and D represent a spring and damper, respectively.
-            K_f = jnp.where(K < 0, -K / ξ_max**2, 1 / (ξ_max * Ω * ζ) ** 2)
-            D_f = jnp.where(D < 0, -D / ξ_max, 2 / (ξ_max * Ω))
+            # If the user specifies K and D and they are negative, the computed `a_ref`
+            # becomes something more similar to a classic Baumgarte regularization.
+            K = jnp.where(K < 0, -K / ξ_max**2, K)
+            D = jnp.where(D < 0, -D / ξ_max, D)
 
-            a_ref = -jnp.atleast_1d(D_f * velocity + K_f * imp * position)
+            # Compute the reference acceleration.
+            a_ref = -(D * vel + K * ξ * pos)
 
-            return imp, a_ref, jnp.atleast_1d(K_f), jnp.atleast_1d(D_f)
+            return ξ, a_ref, K, D
 
-        def _compute_row(
+        def compute_row(
             *,
-            link_idx: jtp.Float,
-            penetration: jtp.Array,
-            velocity: jtp.Array,
-        ) -> tuple[jtp.Array, jtp.Array]:
+            link_idx: jtp.Int,
+            pos: jtp.Vector,
+            vel: jtp.Vector,
+        ) -> tuple[jtp.Vector, jtp.Matrix, jtp.Float, jtp.Float]:
 
             # Compute the reference acceleration.
-            ξ, a_ref, K, D = _imp_aref(
-                penetration=penetration,
-                velocity=velocity,
+            ξ, a_ref, K, D = imp_aref(
+                pos=pos,
+                vel=vel,
             )
 
-            # Compute the regularization terms.
+            # Compute the regularization term.
             R = (
                 (2 * μ**2 * (1 - ξ) / (ξ + 1e-12))
                 * (1 + μ**2)
-                @ jnp.linalg.inv(M_L[link_idx, :3, :3])
+                * (1 / jnp.diag(M_L[link_idx, 0:3, 0:3]).mean())
```

Comment:
@flferretti I suspect that the usage of the spatial inertia of the link $\mathbb{M}\_L$ is wrong here. Accordingly to the following resources, it seems to me that Mujoco initializes it to $\text{diag}(J^\top_{W,C} M^{-1} J_{W,C})[0:3, 0:3]$ computed at $\mathbf{q}_0$ (default configuration from the URDF). It's not a big deal now, let's keep in mind to validate it as soon as possible.

- https://github.com/google/brax/blob/865d974c0285d345425d76df30c38614c828c8c8/brax/generalized/constraint.py
- https://github.com/google/brax/blob/865d974c0285d345425d76df30c38614c828c8c8/brax/base.py#L316C5-L316C14
- https://github.com/google-deepmind/mujoco/blob/5e989d18c349e914fb82ae937006ff4832b011ce/src/engine/engine_setconst.c#L87-L100

Reply:
Delayed to a next PR.